bentoml/_internal/frameworks/xgboost.py

```diff
@@ -162,120 +125,84 @@ def save(
-    *,
-    booster_params: t.Optional[t.Dict[str, t.Union[str, int]]] = None,
-    name: t.Optional[str] = None,
-) -> "_XgBoostRunner":
+def get_runnable(
+    bento_model: bentoml.Model,
+) -> t.Type[bentoml.Runnable]:
+    """
+    Private API: use :obj:`~bentoml.Model.to_runnable` instead.
     """
-    Runner represents a unit of serving logic that can be scaled horizontally to
-    maximize throughput. `bentoml.xgboost.load_runner` implements a Runner class that
-    wrap around a Xgboost booster model, which optimize it for the BentoML runtime.
-
-    Args:
-        tag (:code:`Union[str, Tag]`):
-            Tag of a saved model in BentoML local modelstore.
-        predict_fn_name (:code:`str`, default to :code:`predict`):
-            Options for inference functions. If you want to use `run`
-            or `run_batch` in a thread context then use `inplace_predict`.
-            Otherwise, `predict` are the de facto functions.
-        booster_params (:code:`t.Dict[str, t.Union[str, int]]`, default to :code:`None`):
-            Parameters for boosters. Refers to `Parameters docs <https://xgboost.readthedocs.io/en/latest/parameter.html>`_ for more information.
-
-    Returns:
-        :obj:`~bentoml._internal.runner.Runner`: Runner instances for :mod:`bentoml.xgboost` model
-
-    Examples:
-
-    .. code-block:: python
 
-        import xgboost as xgb
-        import bentoml.xgboost
-        import pandas as pd
+    class XGBoostRunnable(bentoml.Runnable):
+        SUPPORT_NVIDIA_GPU = True
+        SUPPORT_CPU_MULTI_THREADING = True
+
+        def __init__(self):
+            super().__init__()
+            self.model = load_model(bento_model)
+
+            # check for resources
+            available_gpus = os.getenv("NVIDIA_VISIBLE_DEVICES")
+            if available_gpus is not None and available_gpus != "":
+                self.model.set_param({"predictor": "gpu_predictor", "gpu_id": 0})  # type: ignore (incomplete XGBoost types)
+            else:
+                nthreads = os.getenv("OMP_NUM_THREADS")
+                if nthreads is not None and nthreads != "":
+                    nthreads = max(int(nthreads), 1)
+                else:
+                    nthreads = 1
+                self.model.set_param({"predictor": "cpu_predictor", "nthread": nthreads})  # type: ignore (incomplete XGBoost types)
+
+            self.predict_fns: dict[str, t.Callable[..., t.Any]] = {}
+            for method_name in bento_model.info.signatures:
+                self.predict_fns[method_name] = getattr(self.model, method_name)
+
+    for method_name, options in bento_model.info.signatures.items():
+
+        def _run(
+            self: XGBoostRunnable,
+            input_data: ext.NpNDArray | ext.PdDataFrame | xgb.DMatrix,
+        ) -> ext.NpNDArray:
+            if not isinstance(input_data, xgb.DMatrix):
+                input_data = xgb.DMatrix(input_data)
+
+            res = self.predict_fns[method_name](input_data)
```

Comment:
@larme found an issue with this implementation, where as `method_name` is always the last item in the list, due to it being bound to the loop scope variable. We can do something like `predict_fn = self.predict_fns[method_name]` to the beginning of this for loop, to avoid this issue. @sauyon could you help fix that in a follow up PR?

Reply:
Oh, python captures by reference 🤦. Well, some more stuff that should be going into my integration tests...